File: src/freelm/providers/google.py
"""Google AI Studio (Gemini) via its OpenAI-compatible endpoint.

Base: https://generativelanguage.googleapis.com/v1beta/openai
Auth: Authorization: Bearer <AI Studio API key>

Free tier rpm/rpd are per-model and change often; values below are conservative
defaults for flash-class models. Tier 1 (billing enabled) lifts them sharply.
"""
from __future__ import annotations

from typing import Any, Dict

from ..registry import ModelSpec
from .base import Provider


class GoogleAIStudio(Provider):
    name = "google"
    base_url = "https://generativelanguage.googleapis.com/v1beta/openai"

    TIERS: Dict[str, Dict[str, Any]] = {
        "free": {"rpm": 15, "rpd": 1500},
        "tier1": {"rpm": 2000, "rpd": None},
    }

    # Gemini 1.5 is retired for new projects; 2.5 flash family is the current
    # free-tier workhorse, with 2.0 flash kept as a fallback (2026-06).
    DEFAULT_MODELS = [
        ModelSpec("gemini-2.5-flash", ("chat", "fast", "large"), ctx=1000000),
        ModelSpec("gemini-2.5-flash-lite", ("chat", "fast", "small"), ctx=1000000),
        ModelSpec("gemini-2.0-flash", ("chat", "fast"), ctx=1000000),
    ]


# Friendly alias
Gemini = GoogleAIStudio

File: js/src/providers/google.ts
import { modelSpec, ModelSpec } from "../registry.js";
import { Provider, TierLimit } from "./base.js";

/** Google AI Studio (Gemini) via its OpenAI-compatible endpoint. */
export class GoogleAIStudio extends Provider {
  static providerName = "google";
  static baseUrl = "https://generativelanguage.googleapis.com/v1beta/openai";
  static tiers: Record<string, TierLimit> = {
    free: { rpm: 15, rpd: 1500 },
    tier1: { rpm: 2000, rpd: null },
  };
  // Gemini 1.5 is retired for new projects; 2.5 flash family is the current
  // free-tier workhorse, with 2.0 flash kept as a fallback (2026-06).
  // 2.5-flash *thinks by default* and can spend a small max_tokens budget
  // entirely on reasoning (empty text, finish_reason=length) — so the
  // non-thinking lite leads for `auto`, and flash is tagged "reasoning".
  static defaultModels: ModelSpec[] = [
    modelSpec("gemini-2.5-flash-lite", ["chat", "fast", "small"], 1000000),
    modelSpec("gemini-2.5-flash", ["chat", "fast", "large", "reasoning"], 1000000),
    modelSpec("gemini-2.0-flash", ["chat", "fast"], 1000000),
  ];
}

export { GoogleAIStudio as Gemini };
